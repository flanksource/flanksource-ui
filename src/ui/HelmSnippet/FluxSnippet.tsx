import { JSONViewer } from "@flanksource-ui/ui/Code/JSONViewer";
import { useMemo } from "react";
import { ChartData } from "./HelmInstallationSnippets";
import { stringify } from "yaml";
function generateFluxManifests(charts: ChartData[]) {
  const manifests = charts.flatMap((chart) => {
    const documents = [];

    if (chart.createNamespace) {
      documents.push({
        apiVersion: "v1",
        kind: "Namespace",
        metadata: {
          name: chart.namespace
        }
      });
    }

    if (chart.createRepo) {
      documents.push({
        apiVersion: "source.toolkit.fluxcd.io/v1",
        kind: "HelmRepository",
        metadata: {
          name: chart.repoName,
          namespace: chart.namespace
        },
        spec: {
          interval: "5m",
          url: chart.chartUrl
        }
      });
    }

    documents.push({
      apiVersion: "helm.toolkit.fluxcd.io/v2",
      kind: "HelmRelease",
      metadata: {
        name: chart.chart,
        namespace: chart.namespace
      },
      spec: {
        interval: "5m",
        timeout: "10m",
        chart: {
          spec: {
            chart: chart.chart,
            sourceRef: {
              kind: "HelmRepository",
              name: chart.repoName,
              namespace: chart.namespace
            },
            interval: "5m0s"
          }
        },
        values: chart.values
      }
    });

    return documents;
  });

  return manifests.map((doc) => stringify(doc)).join("\n---\n");
}

type Props = {
  data: ChartData[];
};

export default function FluxSnippet({ data }: Props) {
  const yaml = useMemo(() => {
    return generateFluxManifests(
      data.map((data) => ({
        ...data,
        chartUrl: data?.chartUrl ?? "https://flanksource.github.io/charts",
        valueFile: data?.valueFile
          ? // Indent the valueFile content
            data?.valueFile?.replace(/^/gm, "    ")
          : undefined
      }))
    );
  }, [data]);
  return (
    <div className="flex max-h-[30rem] flex-1 flex-col gap-4 overflow-y-auto rounded-md border border-gray-200 p-2 px-4">
      <JSONViewer code={yaml} format="yaml" showLineNo />
    </div>
  );
}
