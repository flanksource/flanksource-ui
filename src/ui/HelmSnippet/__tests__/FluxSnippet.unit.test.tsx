import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import FluxSnippet from "../FluxSnippet";
import { mockInput, mockInputWithKubOptions } from "./mocks/mocks";

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn()
}));

const writeText = jest.fn();

Object.assign(navigator, {
  clipboard: {
    writeText
  }
});

describe("InstallAgentModal", () => {
  it("renders the Helm repository installation command", async () => {
    render(<FluxSnippet data={[mockInput]} />);

    const btn = screen.getByTitle(/Copy to clipboard/i);

    fireEvent.click(btn);

    await waitFor(() => {
      expect(writeText).toHaveBeenCalled();
    });
    expect(writeText.mock["calls"][0][0]).toMatchSnapshot();
  });

  it("renders the Helm repository installation command with kube command", async () => {
    render(<FluxSnippet data={mockInputWithKubOptions} />);
    const btn = screen.getByTitle(/Copy to clipboard/i);

    fireEvent.click(btn);

    await waitFor(() => {
      expect(writeText).toHaveBeenCalled();
    });
    expect(writeText.mock["calls"][0][0]).toMatchSnapshot();
  });
});
