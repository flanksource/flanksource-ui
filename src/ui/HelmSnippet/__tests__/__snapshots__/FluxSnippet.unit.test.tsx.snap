// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallAgentModal renders the Helm repository installation command 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: mission-control-agent

---
apiVersion: source.toolkit.fluxcd.io/v1
kind: HelmRepository
metadata:
  name: flanksource
  namespace: mission-control-agent
spec:
  interval: 5m
  url: https://flanksource.github.io/charts

---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: mission-control-agent
  namespace: mission-control-agent
spec:
  interval: 5m
  timeout: 10m
  chart:
    spec:
      chart: mission-control-agent
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
      interval: 5m0s
  values:
    upstream:
      createSecret: "true"
      host: http://localhost:3000
      username: token
      password: password
      agentName: test-new-agent-instructions
    pushTelemetry:
      enabled: "true"
      topologyName: incident-commander.demo.aws.flanksource.com-test-new-agent-instructions
"
`;

exports[`InstallAgentModal renders the Helm repository installation command with kube command 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name: mission-control-agent

---
apiVersion: source.toolkit.fluxcd.io/v1
kind: HelmRepository
metadata:
  name: flanksource
  namespace: mission-control-agent
spec:
  interval: 5m
  url: https://flanksource.github.io/charts

---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: mission-control-agent
  namespace: mission-control-agent
spec:
  interval: 5m
  timeout: 10m
  chart:
    spec:
      chart: mission-control-agent
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
      interval: 5m0s
  values:
    upstream:
      createSecret: "true"
      host: http://localhost:3000
      username: token
      password: password
      agentName: test-new-agent-instructions
    pushTelemetry:
      enabled: "true"
      topologyName: incident-commander.demo.aws.flanksource.com-test-new-agent-instructions

---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: mission-control-kubernetes
  namespace: mission-control-agent
spec:
  interval: 5m
  timeout: 10m
  chart:
    spec:
      chart: mission-control-kubernetes
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
      interval: 5m0s
  values:
    clusterName: test-new-agent-instructions2
    scraper:
      schedule: "@every 31m"
"
`;
